File: speckit-chat-plugin/src/main/kotlin/com/speckit/plugin/tools/TasksParser.kt
package com.speckit.plugin.tools

import java.io.File

data class TaskItem(
    val id: String,
    val checked: Boolean,
    val parallel: Boolean,
    val story: String?,
    val description: String,
    val rawLine: String,
    val lineNumber: Int
)

data class TaskPhase(
    val number: Int,
    val name: String,
    val priority: String?,
    val isMvp: Boolean,
    val goal: String?,
    val tasks: List<TaskItem>
)

data class TasksFile(
    val featureName: String,
    val phases: List<TaskPhase>,
    val file: File
) {
    val totalTasks get() = phases.sumOf { it.tasks.size }
    val completedTasks get() = phases.sumOf { p -> p.tasks.count { it.checked } }
}

object TasksParser {

    private val featureNamePattern = Regex("""^#\s+Tasks:\s+(.+)""")
    private val phasePattern = Regex("""^##\s+Phase\s+(\d+):\s+(.+)""")
    private val priorityPattern = Regex("""\(Priority:\s*(P\d)\)""")
    private val goalPattern = Regex("""^\*\*Goal\*\*:\s*(.+)""")
    private val taskPattern = Regex("""^-\s+\[([ xX])]\s+(T\d{3,4})\s*(\[P])?\s*(\[US\d+])?\s*(.+)$""")

    fun parse(file: File): TasksFile? {
        if (!file.isFile) return null
        val lines = file.readLines()
        if (lines.isEmpty()) return null

        var featureName = ""
        val phases = mutableListOf<TaskPhase>()
        var currentPhaseNumber = 0
        var currentPhaseName = ""
        var currentPriority: String? = null
        var currentIsMvp = false
        var currentGoal: String? = null
        var currentTasks = mutableListOf<TaskItem>()

        fun flushPhase() {
            if (currentPhaseNumber > 0) {
                phases.add(TaskPhase(currentPhaseNumber, currentPhaseName, currentPriority, currentIsMvp, currentGoal, currentTasks.toList()))
            }
        }

        for ((index, line) in lines.withIndex()) {
            val trimmed = line.trimStart()

            val featureMatch = featureNamePattern.find(trimmed)
            if (featureMatch != null) {
                featureName = featureMatch.groupValues[1].trim()
                continue
            }

            val phaseMatch = phasePattern.find(trimmed)
            if (phaseMatch != null) {
                flushPhase()
                currentPhaseNumber = phaseMatch.groupValues[1].toInt()
                val headerText = phaseMatch.groupValues[2].trim()
                currentPriority = priorityPattern.find(headerText)?.groupValues?.get(1)
                currentIsMvp = headerText.contains("MVP", ignoreCase = true)
                // Strip priority and MVP markers from the name
                currentPhaseName = headerText
                    .replace(priorityPattern, "")
                    .replace(Regex("""MVP""", RegexOption.IGNORE_CASE), "")
                    .replace(Regex("""\(\s*,?\s*\)"""), "")  // leftover empty parens
                    .trim()
                    .trimEnd(',', ' ')
                currentGoal = null
                currentTasks = mutableListOf()
                continue
            }

            val goalMatch = goalPattern.find(trimmed)
            if (goalMatch != null && currentPhaseNumber > 0) {
                currentGoal = goalMatch.groupValues[1].trim()
                continue
            }

            val taskMatch = taskPattern.find(trimmed)
            if (taskMatch != null && currentPhaseNumber > 0) {
                val checked = taskMatch.groupValues[1].lowercase() == "x"
                val id = taskMatch.groupValues[2]
                val parallel = taskMatch.groupValues[3].isNotEmpty()
                val story = taskMatch.groupValues[4].takeIf { it.isNotEmpty() }
                    ?.removeSurrounding("[", "]")
                val description = taskMatch.groupValues[5].trim()
                currentTasks.add(TaskItem(id, checked, parallel, story, description, line, index + 1))
            }
        }
        flushPhase()

        if (phases.isEmpty()) return null
        return TasksFile(featureName, phases, file)
    }
}

File: speckit-chat-vscode/src/tools/tasks-parser.ts
import * as fs from 'fs';

export interface TaskItem {
    id: string;
    checked: boolean;
    parallel: boolean;
    story: string | undefined;
    description: string;
    rawLine: string;
    lineNumber: number;
}

export interface TaskPhase {
    number: number;
    name: string;
    priority: string | undefined;
    isMvp: boolean;
    goal: string | undefined;
    tasks: TaskItem[];
}

export interface TasksFile {
    featureName: string;
    phases: TaskPhase[];
    filePath: string;
    totalTasks: number;
    completedTasks: number;
}

const featureNamePattern = /^#\s+Tasks:\s+(.+)/;
const phasePattern = /^##\s+Phase\s+(\d+):\s+(.+)/;
const priorityPattern = /\(Priority:\s*(P\d)\)/;
const goalPattern = /^\*\*Goal\*\*:\s*(.+)/;
const taskPattern = /^-\s+\[([ xX])]\s+(T\d{3,4})\s*(\[P])?\s*(\[US\d+])?\s*(.+)$/;

export function parse(filePath: string): TasksFile | undefined {
    try {
        const stat = fs.statSync(filePath);
        if (!stat.isFile()) { return undefined; }
    } catch { return undefined; }
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n');
    if (lines.length === 0) { return undefined; }

    let featureName = '';
    const phases: TaskPhase[] = [];
    let currentPhaseNumber = 0;
    let currentPhaseName = '';
    let currentPriority: string | undefined;
    let currentIsMvp = false;
    let currentGoal: string | undefined;
    let currentTasks: TaskItem[] = [];

    function flushPhase(): void {
        if (currentPhaseNumber > 0) {
            phases.push({
                number: currentPhaseNumber,
                name: currentPhaseName,
                priority: currentPriority,
                isMvp: currentIsMvp,
                goal: currentGoal,
                tasks: [...currentTasks],
            });
        }
    }

    for (let i = 0; i < lines.length; i++) {
        const trimmed = lines[i].trimStart();

        const featureMatch = featureNamePattern.exec(trimmed);
        if (featureMatch) {
            featureName = featureMatch[1].trim();
            continue;
        }

        const phaseMatch = phasePattern.exec(trimmed);
        if (phaseMatch) {
            flushPhase();
            currentPhaseNumber = parseInt(phaseMatch[1], 10);
            const headerText = phaseMatch[2].trim();
            const pm = priorityPattern.exec(headerText);
            currentPriority = pm ? pm[1] : undefined;
            currentIsMvp = /MVP/i.test(headerText);
            currentPhaseName = headerText
                .replace(priorityPattern, '')
                .replace(/MVP/i, '')
                .replace(/\(\s*,?\s*\)/, '')
                .trim()
                .replace(/[, ]+$/, '');
            currentGoal = undefined;
            currentTasks = [];
            continue;
        }

        const goalMatch = goalPattern.exec(trimmed);
        if (goalMatch && currentPhaseNumber > 0) {
            currentGoal = goalMatch[1].trim();
            continue;
        }

        const taskMatch = taskPattern.exec(trimmed);
        if (taskMatch && currentPhaseNumber > 0) {
            const checked = taskMatch[1].toLowerCase() === 'x';
            const id = taskMatch[2];
            const parallel = !!taskMatch[3];
            let story: string | undefined = taskMatch[4] || undefined;
            if (story) { story = story.slice(1, -1); } // Remove surrounding [ ]
            const description = taskMatch[5].trim();
            currentTasks.push({
                id, checked, parallel, story, description,
                rawLine: lines[i], lineNumber: i + 1,
            });
        }
    }
    flushPhase();

    if (phases.length === 0) { return undefined; }

    const totalTasks = phases.reduce((sum, p) => sum + p.tasks.length, 0);
    const completedTasks = phases.reduce((sum, p) => sum + p.tasks.filter(t => t.checked).length, 0);

    return { featureName, phases, filePath, totalTasks, completedTasks };
}
